ledger-api/grpc-definitions/com/daml/ledger/api/v1/commands.proto

```diff
@@ -95,6 +127,38 @@ message Commands {
   // to read contract data on behalf of each of the given parties.
   // Optional
   repeated string read_as = 13;
+
+  // A unique identifier to distinguish completions for different submissions with the same change ID.
+  // Typically a random UUID. Applications are expected to use a different UUID for each retry of a submission
+  // with the same change ID.
+  // Must be a valid LedgerString (as described in ``value.proto``).
+  //
+  // If omitted, the participant or Daml ledger may set a value of their choice.
+  // Optional
+  string submission_id = 17;
+
+  // Applications can rank their submissions with the same change ID.
+  // This allows the participants and Daml ledger to detect when one submission has been overtaken by a later resubmission.
+  // The participant and Daml ledger provide the following rank guarantee:
+  // Let there be two submissions for the same change ID with submission ranks `rank1 <= rank2` and submission IDs `sid1` and `sid2`.
+  // Suppose that the submission `sid2` has completed with a definite answer on the completion stream at offset `off2`.
+  // If there is no completion for the submission `sid1` before `off2`, then the submission `sid1` will not result
+  // in a definite answer or successful completion. So if a result ever reaches the application,
+  // the result will be a rejection that is not marked as a definite answer.
+  //
+  // The submission rank is a completion offset.
+  // It must be no earlier than the offset that defines the start of the deduplication period.
+  // Otherwise, the submission is rejected with INVALID_ARGUMENT.
+  // Must be a valid LedgerString (as described in ``value.proto``).
+  //
+  // The participant or Daml ledger may reject the submission with ``OUT_OF_RANGE``
+  // if the submission rank offset is higher than the completion end.
+  // It may reject the submission with NOT_FOUND
+  // if the submission rank is too low (e.g., lower than the the completion begin).
+  //
+  // If omitted, the participant defaults the submission rank to the completion end.
```

Comment:
I've added some changes, but we should consider of whether we want to remove these side-constraints completely; and assume people will discover them either by reading `error.proto` or by encountering the error and then acting on it.